libraries/js/text/regex.effekt

```diff
@@ -1,33 +1,186 @@
+ * val text = "Hello 123 World 456"
+ * while (true) {
+ *   exec(rx, text) match = {
+ *      Some(m) =>
+ *          println("content: " ++ m.content ++ "range: " ++ m.start.show() ++ ", " ++ m.end.show()))
+ *      None() => break
+ * }
+ * ```
+ */
+def exec(reg: Regex, str: String): Option[Match] = {
+    def toGroup(g: GroupMatch): Group = Group(
+        content(g),
+        zip(undefinedToOption(start(g)), undefinedToOption(end(g))) { (s, e) => Range(s, e)}
+    )
+
+    undefinedToOption(reg.unsafeExec(str)) match {
+        case None() => None()
+        case Some(v) =>
+        val fullMatch = fullMatch(v)
+        Some(Match(
+            fullMatch.content,
+            fullMatch.start,
+            fullMatch.end,
+            v.subGroups.toList().map{toGroup}
+        ))
+    }
+}
+
+// internal representation
+extern type GroupMatch
+extern pure def content(g: GroupMatch): String = js "${g}.content"
+extern pure def start(g: GroupMatch): Int = js "${g}.start"
+extern pure def end(g: GroupMatch): Int = js "${g}.end"
+
 extern type RegexMatch
-extern pure def matched(r: RegexMatch): String = js "${r}.matched"
-extern pure def index(r: RegexMatch): Int = js "${r}.index"
+extern pure def fullMatch(r: RegexMatch): GroupMatch = js "${r}.fullMatch"
+extern pure def subGroups(r: RegexMatch): Array[GroupMatch] = js "${r}.groups"
 
 extern js """
 function regex$exec(reg, str) {
-  var res = reg.exec(str);
-  if (res === null) { return undefined }
-  else { return { matched: res[0], index: res.index } }
+  var match = reg.exec(str);
+  if (match === null)  return undefined
+
+  var fullMatch = { content: match[0], start: match.index, end: reg.lastIndex };
+  var groups = [];
+  for (var i = 1; i < match.length; i++) {
+    if (match[i] !== undefined) {
+        if (match.indices !== undefined && match.indices[i] !== undefined) {
+            groups.push({
+                content: match[i],
+                start: match.indices[i][0],
+                end: match.indices[i][1]
+            });
+        } else {
+            groups.push({
+                content: match[i],
+            });
+        }
+    }else {
+        // Push an empty string if an intermediate group is undefined
+        groups.push({content: ""});}
+  }
+
+  return { fullMatch: fullMatch, groups: groups };
 }
+
 """
 
 // internals
 extern io def unsafeExec(reg: Regex, str: String): RegexMatch =
   js "regex$exec(${reg}, ${str})"
+
+
+// Debug show
+def debug(m: Match): String = {
+    val start = m.start.show()
+    val end = m.end.show()
+    val groups = m.groups.map{it => it.debug()}.join("\n")
+
+    "matched: " ++ m.content ++
+    "\nfrom: (" ++ start ++ ", " ++ end ++ ")" ++
+    "\nGroups: \n" ++ groups
+}
+
+def debug(g: Group): String = {
+    val range = g.index.map{ it => it.debug() }.getOrElse{"None"}
+
+    g.content ++  "\nfrom: " ++ range
+}
+
+def debug(r: Range): String = {
+    "(" ++ r.start.show() ++ ", " ++ r.end.show() ++ ")"
+}
```

Comment:
I think you could even use `show` here instead of debug :)
We currently use it for a more debug-ey output anyways.

By the way, you should be able to do `["matched: " ++ m.content, "from: (" ++ start ...].join("\n")` or something similar via [`string::join`](https://github.com/effekt-lang/effekt/blob/1a1d34199e2edce3a4e7b5ea70209feb5d618cf7/libraries/common/string.effekt#L97-L106)